Partner Experience: revamp ingestion docs and guidelines (#168)
stellar/stellar-docs: docs/run-platform-server/ingestion-filtering.mdx

order: 46
---

## Overview

> Ingestion Filtering enables Horizon operators to drastically reduce storage footprint of the historical data on Horizon DB by whitelisting Assets and/or Accounts that are relevant to their operations.

Reviewer: ```suggestion
Ingestion Filtering enables Horizon operators to drastically reduce storage footprint of the historical data in the Horizon database by whitelisting Assets and/or Accounts that are relevant to their operations.
```

Author: resolved in https://github.com/stellar/stellar-docs/pull/168/commits/7f3553bb9c0f5667d8d9aeda841b3f9c454eebc6

> Ingestion Filtering enables Horizon operators to drastically reduce storage footprint of the historical data on Horizon DB by whitelisting Assets and/or Accounts that are relevant to their operations.

Reviewer 2: I know It it is already in our existing doc but consider changing it to  
more industry preferred term like "allowlisting".

Reviewer 3: +1, it shows up in a few other places in here as well

### Why is it useful:

> Previously, the only way to limit data storage is by limiting the temporaral range of history via rolling retention (ie: last 30 days). Filtering feature allows users to store a longer historical timeframe on the Horizon database for only whitelisted assets, accounts and their related historical entities(transactions, operations, trades, etc)

Reviewer: ```suggestion
Previously, the only way to limit data storage was by limiting the temporal range of history via rolling retention (e.g. the last 30 days). The filtering feature allows users to store a longer historical timeframe in the Horizon database for only whitelisted assets, accounts, and their related historical entities (transactions, operations, trades, etc.).
```

For further context, running a non-filtered `full` history Horizon instance currently takes ~ 25TB of disk space (as of June 2023) with storage growing at a rate of ~ 1TB / month. As a benchmark, filtering by even 100 of the most active accounts and assets reduces storage by over 90%. For the majority of applications which are interested in an even more limited set of assets and accounts, storage savings should be well over 99%. Other benefits are reducing operating costs for maintaining storage, improved DB health metrics and query performance.

> ### How does it work:

Author: yes, good points, I added a bunch more on this - https://github.com/stellar/stellar-docs/pull/168/commits/d4ae7d84711f4d5cd049461249136651f73cb019

> This feature operates by accepting only ledger transactions that match to a filter rule criteria when persisting the transactions and operations to historical tables on Horizon database at ingestion time, the rest that done't match are skipped and not stored on database.

Reviewer: ```suggestion
This feature operates by accepting only ledger transactions that match a filter rule when persisting the transactions and operations to historical tables in the Horizon database at ingestion time, any entries that don't match are skipped and not stored on database.
```

> The feature does not filter the storage of current network ledger state in Horizon database as that is required for referential integrity of entity identity in sql database overall, however db storage for current state is relatively extremely lower than historical concerns and does not increase exponentially, in the order of a few GB's total.

Reviewer 3: ```suggestion
Note that this filtering applies only to historical data, and does not affect current state data stored in Horizon. However, current state data consumes a relatively small amount of the overall storage capacity.
```

Filter rules can whitelist ingestion by the following supported entities:

- Account id
- Asset id (canonical)

> Given that all transactions related to the white listed entities are included, all historical time series data related to those transactions are saved in horizon's history db, including transaction itself, all operations in the transaction and references to any ancillary entities from operations.

Reviewer: ```suggestion
Given that all transactions related to the white listed entities are included, all historical time series data related to those transactions are saved in horizon's history db, including transaction itself, all operations in the transaction, and references to any ancillary entities from operations.
```

oxford comma gang

## Configuration:

> Filtering is enabled by default, however, no filter rules are included by default, meaning effectively no filtering of ingested data happens by default. To start filtering ingestion:

Reviewer: suggestion: simplify wording

```suggestion
Filtering is enabled by default but with no filter rules, which effectively means no filtering of ingested data occurs. To start filtering ingestion:
```

> - enable Horizon admin port with environmental configuration parameter `ADMIN_PORT=XXXXX`, this will allow you to access the port.
> - define filter whitelists. submit Admin HTTP API requests to view and update the filter rules:

Reviewer 3: ```suggestion
- enable Horizon admin port with the configuration parameter `ADMIN_PORT=XXXXX`, this will allow you to access the port
- define filter allowlists
- submit Admin HTTP API requests to view and update the filter rules
```

>   Refer to the [Horizon Admin API Docs](https://github.com/stellar/go/blob/horizon-v2.18.0/services/horizon/internal/httpx/static/admin_oapi.yml) which are also published on Horizon instance as Open API 3.0 doc on the Admin Port at `http://localhost:<admin_port>/`. You can paste the contents from that url into any OAPI tool such as [Swagger](https://editor.swagger.io/) which will render a visual explorer of the API endpoints. Follow details and example request/response payloads for these filter rule endpoints:

Reviewer: prob should update the tag here

Author: nice catch, thanks!

  ```
  /ingestion/filters/account
  /ingestion/filters/asset
  ```

### Gap fill on filtered historical data:

> If new Assets or Accounts are added to the whitelist rules and you would like to pull in its missing historical data, which would have been dropped earlie, reingestion can be run. The Reingestion process is idempotent and will re-ingest the data from the designated historical ledger range and `upsert` to Horizon historical data, i.e. overwrite or insert new data if not already on current database.

Reviewer: ```suggestion
If new Assets or Accounts are added to the whitelist rules and you would like to pull in their missing historical data which would have been dropped earlier, you need to run reingestion. The Reingestion process is idempotent and will re-ingest the data from the designated historical ledger range and `upsert` to Horizon historical data, i.e. overwrite or insert new data not already in the current database.
```

## Overview

Ingestion Filtering enables Horizon operators to drastically reduce the storage footprint of the historical data in the Horizon database by allow-listing Assets and/or Accounts that are relevant to their operations.

### Why is it useful:

Previously, the only way to limit data storage was by limiting the temporal range of history via rolling retention (e.g. the last 30 days). The filtering feature allows users to store a longer historical timeframe in the Horizon database for only whitelisted assets, accounts, and their related historical entities (transactions, operations, trades, etc.).

For further context, running an unfiltered `full` history Horizon instance currently requires over 30TB of disk space (as of June 2023) with storage growing at a rate of about 1TB/month. As a benchmark, filtering by even 100 of the most active accounts and assets reduces storage by over 90%. For the majority of applications which are interested in an even more limited set of assets and accounts, storage savings should be well over 99%. Other benefits include reducing operating costs for maintaining storage, improved DB health metrics and query performance.

### How does it work:

> Filtering feature operates during the ingestion process, **live** or **prior historical ranges**. It tells ingestion process to only accept incoming ledger transactions which match on a filter rule, any transactions which don't match on filter rules are skipped by ingestion and therefore not stored on database.

Reviewer 2: > **live** or **prior historical ranges*

Should this be "live and historical ranges"?

Author: yes, i changed that in a couple places for consistency.

Some key aspects to note about filtering behavior:

- Filtering applies only to ingestion of historical data in the database, it does not affect how ingestion process maintains current state data stored in database, which is the last known ledger entry for each unique entity within accounts, trustlines, liquidity pools, offers. However, current state data consumes a relatively small amount of the overall storage capacity.
- When filter rules are changed, they only apply to active ingestion processes(**live** or **historical ranges**). They don't trigger any retro-active filtering or back-filling of existing historical data on the database.

>   - If you update the filter rules to increase allow-listing of accounts or assets, related transactions will only start to show up in historical database data from **live** ingestion beginning after time the filter rule is updated using the Horizon Admin API. Same applies to **historical range** ingestion, it will only be affected by new filter rules starting at current ledger it was processing within it's configured range at time the filter rules were updated.

Reviewer 2: ```suggestion
  - When the filter rules are updated to include additional accounts or assets in the allow-list, the related transactions from **live** ingestion will only appear in the historical database data once the filter rules have been updated. The same applies to **historical range** ingestion, where the new filter rules will only affect the data from the current ledger within its configured range at the time of the update.
```

Author: ok, pulled that in https://github.com/stellar/stellar-docs/pull/168/commits/a12d1394b8f670192aae67b14513c7ab4b51b349

>   - When updating filter rules with increased allow list coverage, no historical back-filling is done automatically. You can manually backfill the history on database by running a new **historical range** ingestion process for a past ledger range after you have updated the filter rules to achieve that result.

Reviewer 2: ```suggestion
  - Updating the filter rules to include additional accounts or assets does not trigger automatic historical back-filling. To backfill the history in the database, you can manually run a new **historical range** ingestion process for a past ledger range after updating the filter rules.
```

Author: included on https://github.com/stellar/stellar-docs/pull/168/commits/a12d1394b8f670192aae67b14513c7ab4b51b349

>   - If you update filter rules and reduce the allow list coverage by removing some entities, no retro-active purging or filtering of historical data per the reduced scope of filter rules on database is performed. Whatever data is stored on history tables resides for lifetime of database or until `HISTORY_RETENTION_COUNT` is exceeded, and Horizon will purge all historical data for all entites related to older ledgers regardless of any filtering rules.

Reviewer 2: ```suggestion
  - When the filter rules are updated to remove accounts or assets from the allow-list, the historical data in the database will not be retroactively purged or filtered based on the updated rules. The data is stored in the history tables for the lifetime of the database or until the `HISTORY_RETENTION_COUNT` is exceeded. Once the retention limit is reached, Horizon will purge all historical data related to older ledgers, regardless of any filtering rules.
```

Author: https://github.com/stellar/stellar-docs/pull/168/commits/a12d1394b8f670192aae67b14513c7ab4b51b349

  - When the filter rules are updated to remove accounts or assets previously defined on white-list, the historical data in the database will not be retroactively purged or filtered based on the updated rules. The data is stored in the history tables for the lifetime of the database or until the `HISTORY_RETENTION_COUNT` is exceeded. Once the retention limit is reached, Horizon will purge all historical data related to older ledgers, regardless of any filtering rules.
- Filtering will not affect the performance or throughput rate of an ingestion process, it will remain consistent whether filter rules are present or not.

Filter rules define white-lists of the following supported entities:

- Account id
- Asset id (canonical)

Given that all transactions related to the white listed entities are included, all historical time series data related to those transactions are saved in horizon's history db, including transaction itself, all operations in the transaction, and references to any ancillary entities from operations.

## Configuration:

> Filtering is enabled by default but with no filter rules, which effectively means no filtering of ingested data occurs. To start filtering ingestion:

Reviewer 2: I am unsure if by "no filtering of ingested data" means that all data is filtered out (and not stored), or if it means that all data is stored? If it's the latter, considering we only have allow-lists, it seems a bit counterintuitive to me that having no allow-list rules would result in everything allowed and stored.

Author: updated this with more explanation on how at least one filter rule is needed to trigger filtering, if no filter rules then triggers no filtering. related pr on making this effective - https://github.com/stellar/go/pull/4955

## Overview

> Ingestion Filtering enables Horizon operators to drastically reduce the storage footprint of the historical data in the Horizon database by allow-listing Assets and/or Accounts that are relevant to their operations.

Reviewer 2: Although I initially suggested changing 'whitelist' to 'allowlist,' I now realize that the API response already uses the term like this `{"whitelist":["accountid1","accountid2","accountid3"],"enabled":true,"last_modified":1689048762}` So, should we stick to the same terminology in the document as well? I apologize for any extra work this may have caused. Thanks!

Author: ah, nice catch, np, will change that back for consistency, it was just a couple places specific to horizon docs

### Pre-requisites:

> You have installed Horizon with empty database, it has **live** ingestion enabled.

Reviewer: ```suggestion
You have installed Horizon with empty database and it has **live** ingestion enabled.
```

### Steps:

> 1. Configure a filter rule with 4 whitelisted Assets via the Admin API. Also check the `HISTORY_RETENTION_COUNT` and set it to `0` if you don’t want any history purged anymore now that you are filtering, otherwise it will continue to reap all data older than the retention.

Reviewer 3: ```suggestion
1. Configure a filter rule with 4 whitelisted Assets via the Admin API
```
I would just leave this part out. Someone could in theory want to retain limited history *and* use ingestion filtering. And think we should just overall avoid plugging full history use cases

Author: yes, will drop, it's just a compelling side effect of filtering that wanted to highlight it as a way to encourage it's usage moreso.
